### PR TITLE
feat: multiple lg

### LINKED ISSUE #796: Move LG away from common and to one .lg file per dialog
@cwhitten @yochay @boydc2014 @hibrenda - If possible, we should get this slotted for Ignite. 

- [ ] Remove common.lg 
- [ ] Add one .lg file per dialog
- [ ] Verify ability to cross reference/ include other .lg content by dialog name
    e.g. [Include dialog A](dialogA) - here, dialogA should be resolved by the file resolver to map to dialogA's .lg file. 
- [ ] Each dialog can have at most one .lg file.
- [ ] Migration plan: once we have this in: try to look for a template in common.lg, if we find it, serialize it in to the dialog's .lg file.

### LINKED ISSUE #1871: [Inline LG] lsp error info disagrees with diagnostics.
Complete the necessary portions of this template and delete the rest.

## Describe the bug
`monacoEditor` error behavior disagree with diagnostics.

## To Reproduce
1. Create a `sendResponse` step, there is a error in editor, but diagnostics is null.

## Expected behavior
There should not have error in `monacoEditor`.

## Screenshots
image

### PR DESCRIPTION
**Pull request by zhixzhan**
## Description

Add support for multiple lg file.

##### 1. Bot file structure:
Each when create a dialog, lg/lu will also be created. same as deleting.
```
+ ToDoSample
    + common
       - common.lg
    + Main
       - Main.dialog
       - Main.lg
       - Main.lu
```

##### 2. support lg import
```
[import](../common/common.lg) 
 
# bfdactivity-696707
- @{help()}

# bfdactivity-157674()
- Hi! I'm a ToDo bot. Say "add a todo named first" to get started.

``` 
`lgIndexer/parse,check`  now support call with an `importResolverDelegate`.
and default importResolverDelegate implemented in server and client/store

##### 3. visual editor

new created lg, will save to dialog's lg file, not common.lg anymore.

##### 4. auto migration from single-lg bot
auto move generated lg template (like bfdactivity-123456) from common.lg into dialog.lg

## Task Item

close #796 
close #1871 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code refactor (non-breaking change which improve code quality, clean up, add tests, etc)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Doc update (document update)

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have functionally tested my change

## Screenshots

preview

### CODE REVIEW

**Comment by vishwacsena** [comment]
If and when we resurrect this, we should support import with just dialog name. users do not have to know the file persistance. 

```
> Ok to support this but should not be expected to be this.
[import](../common/common.lg) 
> we should support this. I expect that we will support named collection of templates in 'common' in which case those should all be importable via name or possibly common#collectionName notation if we plan to use sections in LG to represent those collections.
[common](common)
[dialog1](dialog1)
```

**Comment by zhixzhan** [comment]
> If and when we resurrect this, we should support import with just dialog name. users do not have to know the file persistance.
> 
> ```
> > Ok to support this but should not be expected to be this.
> [import](../common/common.lg) 
> > we should support this. I expect that we will support named collection of templates in 'common' in which case those should all be importable via name or possibly common#collectionName notation if we plan to use sections in LG to represent those collections.
> [common](common)
> [dialog1](dialog1)
> ```
> 
> @boydc2014 @zhixzhan

OK, it make sense, will update this.

**Comment by boydc2014** [comment]
> > If and when we resurrect this, we should support import with just dialog name. users do not have to know the file persistance.
> > ```
> > > Ok to support this but should not be expected to be this.
> > [import](../common/common.lg) 
> > > we should support this. I expect that we will support named collection of templates in 'common' in which case those should all be importable via name or possibly common#collectionName notation if we plan to use sections in LG to represent those collections.
> > [common](common)
> > [dialog1](dialog1)
> > ```
> > 
> > 
> > @boydc2014 @zhixzhan
> 
> OK, it make sense, will update this.

@vishwacsena  @zhixzhan , perhaps we should make it support `[import](common.lg)` and make this as default in all samples. This is the convention SDK is using today and more clearly says about importing lg.

@zhixzhan in you customized resolver,  you probably want to support common and common.lg both.

**Comment by zhixzhan** [comment]
> > > If and when we resurrect this, we should support import with just dialog name. users do not have to know the file persistance.
> > > ```
> > > > Ok to support this but should not be expected to be this.
> > > [import](../common/common.lg) 
> > > > we should support this. I expect that we will support named collection of templates in 'common' in which case those should all be importable via name or possibly common#collectionName notation if we plan to use sections in LG to represent those collections.
> > > [common](common)
> > > [dialog1](dialog1)
> > > ```
> > > 
> > > 
> > > @boydc2014 @zhixzhan
> > 
> > 
> > OK, it make sense, will update this.
> 
> @vishwacsena @zhixzhan , perhaps we should make it support `[import](common.lg)` and make this as default in all samples. This is the convention SDK is using today and more clearly says about importing lg.
> 
> @zhixzhan in you customized resolver, you probably want to support common and common.lg both.

Sure, now those 3 usage case are supported
```
1. [import](common) 
2. [import](common.lg) 
3. [import](./[path]/common.lg)
 ```
I can update sample bot with case-2 syntax.

File: Composer/packages/server/src/models/bot/botProject.ts [hunk 1]
```diff
@@ -537,6 +589,63 @@ export class BotProject {
...
+          const lineCount = endLineNumber - startLineNumber + 1;
+          const templateText = lineContentArray
+            .splice(startLineNumber - 1, lineCount, ...Array(lineCount))
+            .join(NEWLINE);
+          dialogTemplateTexts.push(templateText);
+        }
+      }
+      const updatedContent =
+        (lgFiles.find(({ id }) => id === dialog.id)?.content || '') + dialogTemplateTexts.join(DUBNEWLINE);
+      await this.updateLgFile(dialog.id, updatedContent);
```
Comment (boydc2014) [inline Composer/packages/server/src/models/bot/botProject.ts]: See here

**Comment by boydc2014** [comment]
One last thing, after some digging into, i finally can repro the warning message in console. Create a bot with the master branch, and use this branch to open it. You will see some errors in console complaining about '\r'.

I believe there is some thing happening in your conversion phrase (by the way, this conversion logic is great), i think perhaps there are some issue in this version of lg parser when doing CRUD, not sure. Please take a look.

**Comment by zhixzhan** [comment]
> One last thing, after some digging into, i finally can repro the warning message in console. Create a bot with the master branch, and use this branch to open it. You will see some errors in console complaining about '\r'.
> 
> I believe there is some thing happening in your conversion phrase (by the way, this conversion logic is great), i think perhaps there are some issue in this version of lg parser when doing CRUD, not sure. Please take a look.

Hi @boydc2014 @cwhitten After some investments, I found two reason cause this error lg.
1. lg Parser do not support file end with '\r', that's why we could only re-pro it in windows, not *nix.
2. In split common.lg, I use '\n' as NEWLINE splitter, but windows NEWLINE is '\r\n'.

To fix 2, I switch to use `import { EOL } from 'os';` which provide by nodejs standard NEWLINE.
In *nix EOL would be '\n'
In windows EOL would be '\r\n'

By doing this, we can ensure Composer would not produce lg file end with `\r`. But for other text editor, it may still be an bug for lg parser.

**Pull request merged**